### PR TITLE
Use full path to script when executing with run command

### LINKED ISSUE #4984: --cwd with relative paths breaks binstubs
**Do you want to request a *feature* or report a *bug*?**
Bugish.

**What is the current behavior?**
When running `yarn --cwd some/rel/path run some-script`, the `some-script` in `some/rel/path/package.json` is invoked with `some/rel/path/node_modules/.bin:...`, meaning the binstubs are not found.

**If the current behavior is a bug, please provide the steps to reproduce.**

```
% mkdir -p yarn-test/sub/dir

% cd yarn-test

% yarn init -y

% yarn add cat-ascii-faces
yarn add v1.3.2
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
[4/4] Building fresh packages...
success Saved lockfile.
success Saved 1 new dependency.
└─ cat-ascii-faces@2.0.0
Done in 0.58s.

% yarn cat-face
yarn run v1.3.2
$ /mnt/c/code/personal/yarn-test/node_modules/.bin/cat-face
~(=^‥^)/
Done in 0.20s.

% cd sub/dir

% yarn cat-face
yarn run v1.3.2
$ /mnt/c/code/personal/yarn-test/node_modules/.bin/cat-face
=＾● ⋏ ●＾=
Done in 0.21s.
```

Working as expected. Now, try using it with `--cwd`:

```
% yarn --cwd ../.. cat-face
yarn run v1.3.2
$ ../../node_modules/.bin/cat-face
/bin/sh: 1: ../../node_modules/.bin/cat-face: not found
error Command failed with exit code 127.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

% cd ../../..

% yarn --cwd yarn-test cat-face
yarn run v1.3.2
$ yarn-test/node_modules/.bin/cat-face
/bin/sh: 1: yarn-test/node_modules/.bin/cat-face: not found
error Command failed with exit code 127.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Note that it is resolving the binstub first, but using a relative path that is no longer valid after the current directory changes. To show the difference to an incorrect binstub:

```
% yarn --cwd yarn-test no-such-binstub
yarn run v1.3.2
error Command "no-such-binstub" not found.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

Easiest to see what's happening with built-in `run env`, using `jq` (should be in most package managers) to parse the path env out of the JSON.

Without --cwd
```
% cd yarn-test

% yarn bin
/mnt/c/code/personal/yarn-test/node_modules/.bin

% PATH=/usr/bin:/bin yarn -s run env | jq -r '.PATH | split(":")[]'
/mnt/c/code/personal/yarn-test/node_modules/.bin
/home/simon/.config/yarn/link/node_modules/.bin
/mnt/c/code/personal/yarn-test/node_modules/.bin
/home/simon/.config/yarn/link/node_modules/.bin
/home/simon/.yarn/bin
/usr/lib/node_modules/npm/bin/node-gyp-bin
/usr/bin/node_modules/npm/bin/node-gyp-bin
/usr/bin
/bin

% cd sub/dir

% yarn --cwd ../.. bin
../../node_modules/.bin

% PATH=/usr/bin:/bin yarn --cwd ../.. -s run env | jq -r '.PATH | split(":")[]'
../../node_modules/.bin
/home/simon/.config/yarn/link/node_modules/.bin
../../node_modules/.bin
/home/simon/.config/yarn/link/node_modules/.bin
/home/simon/.yarn/bin
/usr/lib/node_modules/npm/bin/node-gyp-bin
/usr/bin/node_modules/npm/bin/node-gyp-bin
/usr/bin
/bin
```

Interesting that it doubles the binstub dirs.

**What is the expected behavior?**

`yarn bin` and `PATH` should be absolute.

**Please mention your node.js, yarn and operating system version.**

Node: 6.12.0
Yarn: 1.3.2
OS: Ubuntu 16.10 running on WSL on Windows 10 (Fall Creators Update)

### LINKED ISSUE #5327: offline cache fails with relative --cwd
**Do you want to request a *feature* or report a *bug*?**
Bug.

**What is the current behavior?**
`yarn install --cwd some/relative/path` fails to fill the offline cache directory

whereas...

`yarn install --cwd /some/absolute/path` successfully fills the offline cache directory.

**If the current behavior is a bug, please provide the steps to reproduce.**

```sh
#!/bin/sh

# config
PARENT_PATH="${PWD}/parent"
SUB_PATH="${PARENT_PATH}/sub"
OFFLINE_MIRROR_PATH="${PARENT_PATH}/.yarn-offline-mirror"

# setup the test scenario
mkdir -p "$SUB_PATH"
cd "$SUB_PATH"
yarn init --yes
yarn add semver
rm -rf node_modules

# enable offline mirror
echo yarn-offline-mirror "${OFFLINE_MIRROR_PATH}" >"${PARENT_PATH}/.yarnrc"

# install from within sub directory
yarn install
echo ">> after install from sub directory, offline mirror contains..."
ls "$OFFLINE_MIRROR_PATH"
# as expected, offline mirror has package within it

# clean up
rm -rf "${SUB_PATH}/node_modules"
rm -rf "$OFFLINE_MIRROR_PATH"
echo ">> after cleanup, offline mirror contains..."
ls "$OFFLINE_MIRROR_PATH"

cd "$PARENT_PATH"

yarn install --cwd "$SUB_PATH"
echo ">> after install from parent directory (absolute path), offline mirror contains..."
ls $OFFLINE_MIRROR_PATH
# as expected, offline mirror has package within it

# clean up
rm -rf "${SUB_PATH}/node_modules"
rm -rf "$OFFLINE_MIRROR_PATH"
echo ">> after cleanup, offline mirror contains..."
ls "$OFFLINE_MIRROR_PATH"

yarn install --cwd sub
echo ">> after install from parent directory (relative path), offline mirror contains..."
ls $OFFLINE_MIRROR_PATH
# error, offline mirror has no package within it
```

**What is the expected behavior?**

Relative `--cwd` paths should work identically to absolute `--cwd` paths.

**Please mention your node.js, yarn and operating system version.**

node --version => v8.9.1
yarn --version => 1.3.2
macOS 10.13.3

### PR DESCRIPTION
**Pull request by rally25rs**
fixes #4984 
fixes #5327 

**Summary**

Previously the relative path to the script to run was being passed to the child_process function, plus the cwd was being set. This caused relative paths to be doubled.

For example if the current `process.cwd()` is `/projects/sampleProject/subdir` and you execute `yarn --cwd .. run test` then it would attempt to run the command

> ../node_modules/.bin/test

from

> /projects/sampleProject

which is incorrect.

The fix here was to use `path.resolve` instead of `.join` to convert the script's location to an absolute path.

**Test plan**

Tested manually.

I tried to write a unit test for this, but under the current test framework an absolute dir was always passed to the script execution because `config.cwd` is always set to the absolute path for the test fixture.

Trying to set the `config.cwd` to a relative path causes tests to fail because they become relative to where `yarn run test` was executed (the root of the Yarn project).

I gave up 😢

### CODE REVIEW

File: src/cli/commands/run.js [hunk 1]
```diff
@@ -26,7 +26,7 @@ export async function run(config: Config, reporter: Reporter, flags: Object, arg
   const visitedBinFolders = new Set();
   let pkgCommands = [];
   for (const registry of Object.keys(registries)) {
-    const binFolder = path.join(config.cwd, config.registries[registry].folder, '.bin');
+    const binFolder = path.resolve(config.cwd, config.registries[registry].folder, '.bin');
```
Comment (kaylieEB) [inline src/cli/commands/run.js]: It looks like we use `path.join(config.cwd)` in several places in the code. Would it be worth updating https://github.com/yarnpkg/yarn/blob/master/src/config.js#L358 to be `path.resolve(opts.cwd)` for future issues related to relative cwd path?

Comment (rally25rs) [reply src/cli/commands/run.js]: actually, that's a good idea... @BYK thoughts? If we change it in `config` then `config.cwd` would always be an absolute path. I can't think of any situation where that would be bad.

Comment (BYK) [reply src/cli/commands/run.js]: +1 from me. I'm hoping tests will protect us from any unforeseen consequences.

Comment (rally25rs) [reply src/cli/commands/run.js]: Pushed a new commit that adds a `path.resolve` to `config.cwd` to ensure it is always an absolute path.

**Comment by kaylieEB** [comment]
looks good to merge, but can't due to unrelated failure :(

**Comment by mark-buer** [comment]
What is the status of this PR?

Merging it will also fix #5327.

**Comment by rally25rs** [comment]
I forgot about it, honestly... I kicked off a new Appveyor build. Let's see if it passes this time... If so then we can merge it.

**Comment by rally25rs** [comment]
I have no idea why Appveyor won't pass only for node v4. I merged latest `master` into it too. The 1 line change that this PR is shouldn't be breaking anything, especially not causing a timeout error only on node v4.

@BYK thoughts? think we can just merge this? (I can't override the required passing checks)

**Comment by kaylieEB** [comment]
I tried re-kicking the build, I'll merge if everything passes and no one objects :) 🎈

**Comment by simonbuchan** [comment]
Never knew node 4 was so particular about your grammar 😁

**Pull request merged**